src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -1204,19 +1240,214 @@ private async Task PingServer()
             }
         }
 
-        private async Task RunClosedEvent(Func<Exception, Task> closed, Exception closeException)
+        private async Task HandleConnectionClose(ConnectionState connectionState)
         {
-            // Dispatch to the thread pool before we invoke the user callback
-            await AwaitableThreadPool.Yield();
+            // Clear the connectionState field
+            await WaitConnectionLockAsync();
+            try
+            {
+                SafeAssert(ReferenceEquals(_connectionState, connectionState),
+                    "Someone other than ReceiveLoop cleared the connection state!");
+                _connectionState = null;
+            }
+            finally
+            {
+                ReleaseConnectionLock();
+            }
 
+            // Dispose the connection
+            await CloseAsync(connectionState.Connection);
+
+            // Cancel any outstanding invocations within the connection lock
+            connectionState.CancelOutstandingInvocations(connectionState.CloseException);
+
+            if (connectionState.Stopping || _reconnectPolicy != null)
```

Comment:
Should this be `_reconnectPolicy == null`?

Reply:
Yep. As you can tell, this PR is still in the very early stages.

Reply:
Sorry, thanks for humoring my review, I have been a big SignalR fan and have been eagerly watching for this feature throughout the clients. It has also been a great learning experience seeing how it is implemented under the covers. Didn't mean to rush you! Thank you guys for all the hard work.

Reply:
Thanks for the review. You've raised a lot of good points. I just wanted to make sure you knew there's a lot left for me to do.